core/src/main/java/ai/timefold/solver/core/impl/heuristic/move/AbstractUndoMove.java

```diff
@@ -0,0 +1,71 @@
+package ai.timefold.solver.core.impl.heuristic.move;
+
+import java.util.Collection;
+import java.util.Objects;
+
+import ai.timefold.solver.core.api.domain.solution.PlanningSolution;
+import ai.timefold.solver.core.api.score.director.ScoreDirector;
+
+/**
+ * Abstract superclass for {@link Move}, suggested starting point to implement undo moves
+ * when not using {@link AbstractSimplifiedMove}.
+ * Unless raw performance is a concern, consider using {@link AbstractSimplifiedMove} instead.
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ * @see Move
+ */
+public abstract class AbstractUndoMove<Solution_> implements Move<Solution_> {
+
+    protected final Move<Solution_> parentMove;
+
+    protected AbstractUndoMove(Move<Solution_> parentMove) {
+        this.parentMove = Objects.requireNonNull(parentMove);
+    }
+
+    @Override
+    public final boolean isMoveDoable(ScoreDirector<Solution_> scoreDirector) {
+        return true; // Undo moves are always doable; the parent move was already done.
+    }
+
+    @Override
+    public final Move<Solution_> doMove(ScoreDirector<Solution_> scoreDirector) {
+        throw new UnsupportedOperationException("""
```

Comment:
Do we need a text block here?

Reply:
Personally, I think it's better if long exception messages are split in multiple lines,
for readability both in code and in sysout.
And to do that, I think a text block is more natural than `\n`.